internal/s3storage/s3storage.go

```diff
@@ -1,54 +1,149 @@
+	s3client               *s3.S3
+	sqsClient              *sqs.SQS
+	tracer                 tracing.Tracer
+	sqsQueueURL            string
+	sqsQueueARN            string
+	sqsHandlerQuitChannel  chan struct{}
+	sqsHandlerErrorChannel chan error
 }
 
 func New(bucketName string, tracer tracing.Tracer) (*Service, error) {
+	region := "eu-west-1"
 	sess, err := session.NewSession(&aws.Config{
-		Region: aws.String("eu-west-1"),
+		Region: aws.String(region),
 	})
 	if err != nil {
 		return nil, err
 	}
-	// Create a S3 client from just a session.
+
 	s3client := s3.New(sess)
+	sqsClient := sqs.New(sess)
 
 	return &Service{
+		region:     region,
 		bucketName: bucketName,
 		s3client:   s3client,
+		sqsClient:  sqsClient,
 		tracer:     tracer,
 	}, nil
 }
 
+// InitializeSQS set up a subscription from S3 to a SQS queue. The SQS queue events triggers calls to handler.InitializeSQS
+//
+// Note on S3 notifications:
+// There are 2 ways to get S3 notifications into SQS:
+//  - Directly from S3 Notification to SQS Queue
+//  - From S3 Notification to SNS Topic to SQS Queue
+// Using a SNS Topic should be more powerful, since it fx can send to multiple SQS Queues, but it requires more configuration
+// and moving parts. The simpler model should suffice, therefore we connect it directly.k
+func (s *Service) InitializeSQS(handler func(msg string) error) error {
+	// Amazon SQS returns only an error if the request includes attributes whose values differ from those of the existing queue.
+	queue, err := s.sqsClient.CreateQueue(&sqs.CreateQueueInput{
+		QueueName: aws.String(s.sqsQueueName()),
+	})
+	if err != nil {
+		return errors.Wrap(err, "create sqs queue")
+	}
+	log.Infof("Created queue with url: %s", *queue.QueueUrl)
+	s.sqsQueueURL = *queue.QueueUrl
+
+	queueAttributes, err := s.sqsClient.GetQueueAttributes(&sqs.GetQueueAttributesInput{
+		QueueUrl:       queue.QueueUrl,
+		AttributeNames: []*string{aws.String(sqs.QueueAttributeNameQueueArn)},
+	})
+	if err != nil {
+		return errors.Wrap(err, "get sqs queue ARN")
+	}
+	s.sqsQueueARN = *queueAttributes.Attributes[sqs.QueueAttributeNameQueueArn]
+	log.Infof("SQS queue arn acquired: %s", s.sqsQueueARN)
+
+	policy := `{
+		"Version": "2012-10-17",
+		"Id": "` + s.sqsQueueARN + `/SQSDefaultPolicy",
+		"Statement": [
+		 {
+			"Sid": "` + s.bucketName + `-s3-permission",
+			"Effect": "Allow",
+			"Principal": {
+			 "AWS":"*"
+			},
+			"Action": [
+			 "SQS:SendMessage"
+			],
+			"Resource": "` + s.sqsQueueARN + `",
+			"Condition": {
+				 "ArnLike": { "aws:SourceArn": "arn:aws:s3:*:*:` + s.bucketName + `" }
+			}
+		 }
+		]
+	 }`
+
+	_, err = s.sqsClient.SetQueueAttributes(&sqs.SetQueueAttributesInput{
+		QueueUrl: &s.sqsQueueURL,
+		Attributes: aws.StringMap(map[string]string{
+			sqs.QueueAttributeNamePolicy: policy,
+		}),
+	})
+	if err != nil {
+		log.With("policy", policy, "bucketName", s.bucketName, "Error", fmt.Sprintf("%s", err)).Errorf("Failed update policy: %s", err)
```

Comment:
The log package will serialize errors correctly, so no need for `fmt.Sprintf`.
```suggestion
		log.With("policy", policy, "bucketName", s.bucketName, "error", err).Errorf("Failed update policy: %v", err)
```

Reply:
Oh, great! I stole that from https://github.com/lunarway/release-manager/blob/s3-notifications/internal/broker/amqp/consumer.go#L106:L106
but I see my mistake now 👍